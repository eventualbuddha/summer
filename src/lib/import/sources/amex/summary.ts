import { parseAmount } from '$lib/import/parse/money';
import type { Page } from '$lib/import/statement/page';
import { Result } from '@badrap/result';
import { DateTime } from 'luxon';
import { ParseStatementSummaryError } from '../../parse/errors';
import { StatementSummary } from './statement';

export function parseStatementSummary(
	page: Page
): Result<StatementSummary, ParseStatementSummaryError> {
	const closingDateLabel = page.navigator.find('Closing Date');
	const closingDateValue = closingDateLabel?.findRight(/^\d+\/\d+\/\d+$/);

	if (!closingDateValue) {
		return Result.err(new ParseStatementSummaryError(page.pageNumber, 'Missing closing date'));
	}

	const closingDate = DateTime.fromFormat(closingDateValue.text.str.trim(), 'MM/dd/yy');

	if (!closingDate.isValid) {
		return Result.err(
			new ParseStatementSummaryError(
				page.pageNumber,
				`Invalid closing date: ${closingDate.invalidExplanation}`
			)
		);
	}

	const accountHolderLabel = closingDateLabel?.findUp(/\w+/, { alignment: 'left', maxGap: 10 });

	if (!accountHolderLabel) {
		return Result.err(new ParseStatementSummaryError(page.pageNumber, 'Missing account holder'));
	}

	const accountEndingLabel = closingDateLabel?.findDown(/Account Ending/, { alignment: 'left' });
	const accountEndingValue = accountEndingLabel?.findRight(/[-\d]+/);

	if (!accountEndingValue) {
		return Result.err(new ParseStatementSummaryError(page.pageNumber, 'Missing account number'));
	}

	const previousBalanceLabel = page.navigator.find(/Previous Balance/);
	const entries = previousBalanceLabel?.findTable(
		[
			/Previous Balance/,
			/Payments\/Credits/,
			/New Charges/,
			/Fees/,
			/Interest Charged/,
			/New Balance/,
			/Minimum Payment Due/,
			/Credit Limit/,
			/Available Credit/,
			/Cash Advance Limit/,
			/Available Cash/
		],
		[/\d+/, /\d+/, /\d+/, /\d+/, /\d+/, /\d+/, /\d+/, /\d+/, /\d+/, /\d+/, /\d+/],
		{ labelAlignment: 'left', valueAlignment: 'right' }
	);

	if (!entries) {
		return Result.err(new ParseStatementSummaryError(page.pageNumber, 'Missing summary info'));
	}

	const account = accountEndingValue.text.str.trim();
	const accountHolder = accountHolderLabel.text.str.trim();
	const previousBalance = parseAmount(entries[0][1].text.str);
	const paymentsAndCredits = parseAmount(entries[1][1].text.str);
	const newCharges = parseAmount(entries[2][1].text.str);
	const fees = parseAmount(entries[3][1].text.str);
	const interestCharged = parseAmount(entries[4][1].text.str);
	const newBalance = parseAmount(entries[5][1].text.str);
	const minimumPaymentDue = parseAmount(entries[6][1].text.str);
	const creditLimit = parseAmount(entries[7][1].text.str);
	const availableCredit = parseAmount(entries[8][1].text.str);
	const cashAdvanceLimit = parseAmount(entries[9][1].text.str);
	const availableCash = parseAmount(entries[10][1].text.str);

	const result = Result.all([
		previousBalance,
		paymentsAndCredits,
		newCharges,
		fees,
		interestCharged,
		newBalance,
		minimumPaymentDue,
		creditLimit,
		availableCredit,
		cashAdvanceLimit,
		availableCash
	]);

	if (result.isErr) {
		return Result.err(
			new ParseStatementSummaryError(page.pageNumber, 'Failed to parse summary info', result.error)
		);
	}

	return Result.ok(new StatementSummary(closingDate, account, accountHolder, ...result.value));
}
