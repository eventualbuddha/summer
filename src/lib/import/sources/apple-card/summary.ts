import { ParseStatementError } from '$lib/import/parse/errors';
import { MONEY_PATTERN, parseAmount } from '$lib/import/parse/money';
import type { Statement } from '$lib/import/statement/Statement';
import { StatementMetadata } from '$lib/import/StatementMetadata';
import { Result } from '@badrap/result';
import { DateTime } from 'luxon';

const MMM_DD_YYYY_PATTERN = /^\s*\w+ \d{1,2}, \d{4}\s*$/;
const MMM_DD_PATTERN = /^\s*\w+ \d{1,2}(, \d{4})?\s*$/;

export function parseStatementSummary(
	statement: Statement
): Result<StatementSummary, ParseStatementError> {
	const statementHeading = statement.navigator.find('Statement');

	if (!statementHeading) {
		return Result.err(ParseStatementError.MissingLabel('Missing statement heading'));
	}

	const closingDateValue = statementHeading.findDown(MMM_DD_YYYY_PATTERN, {
		maxGap: 50,
		alignment: 'right'
	});

	if (!closingDateValue) {
		return Result.err(
			ParseStatementError.MissingValue('Missing closing date value', {
				searchFromText: statementHeading.text,
				searchDirection: 'down'
			})
		);
	}

	const closingDate = DateTime.fromFormat(closingDateValue.text.str.trim(), 'MMM d, yyyy');

	const startDateValue = closingDateValue.findLeft(MMM_DD_PATTERN);

	if (!startDateValue) {
		return Result.err(
			ParseStatementError.MissingValue('Missing start date value', {
				searchFromText: closingDateValue.text,
				searchDirection: 'left'
			})
		);
	}

	let startDate = DateTime.fromFormat(startDateValue.text.str.trim(), 'MMM d');

	if (!startDate.isValid) {
		startDate = DateTime.fromFormat(startDateValue.text.str.trim(), 'MMM d, yyyy');
	} else {
		startDate = startDate.set({ year: closingDate.year });
	}

	const yourBalanceLabel = statement.navigator.find(/Your \w+ Balance/);

	if (!yourBalanceLabel) {
		return Result.err(ParseStatementError.MissingLabel('Missing balance label'));
	}

	const yourBalanceValue = yourBalanceLabel.findDown(MONEY_PATTERN, {
		maxGap: 50,
		alignment: 'left'
	});

	if (!yourBalanceValue) {
		return Result.err(
			ParseStatementError.MissingValue('Missing balance value', {
				searchFromText: yourBalanceLabel.text,
				searchDirection: 'down'
			})
		);
	}

	const yourBalance = parseAmount(yourBalanceValue.text.str);

	if (yourBalance.isErr) {
		return Result.err(
			ParseStatementError.InvalidValue(`Invalid balance: ${yourBalance.error}`, {
				pageNumber: yourBalanceValue.pageNumber
			})
		);
	}

	const minimumPaymentDueValue = yourBalanceValue.findRight(MONEY_PATTERN);

	if (!minimumPaymentDueValue) {
		return Result.err(
			ParseStatementError.MissingValue('Missing minimum payment due value', {
				searchFromText: yourBalanceLabel.text,
				searchDirection: 'right'
			})
		);
	}

	const paymentDueDateValue = minimumPaymentDueValue.findRight(MMM_DD_YYYY_PATTERN);

	if (!paymentDueDateValue) {
		return Result.err(
			ParseStatementError.MissingValue('Missing payment due date value', {
				searchFromText: minimumPaymentDueValue.text,
				searchDirection: 'right'
			})
		);
	}

	const paymentDueDate = DateTime.fromFormat(paymentDueDateValue.text.str.trim(), 'MMM d, yyyy');

	const previousBalanceLabel = paymentDueDateValue.findDown(/Previous Total Balance/, {
		alignment: 'left'
	});

	if (!previousBalanceLabel) {
		return Result.err(
			ParseStatementError.MissingLabel('Missing previous balance label', {
				searchFromText: paymentDueDateValue.text,
				searchDirection: 'down'
			})
		);
	}

	const previousBalanceValue = previousBalanceLabel.findRight(MONEY_PATTERN);

	if (!previousBalanceValue) {
		return Result.err(
			ParseStatementError.MissingValue('Missing previous balance value', {
				searchFromText: previousBalanceLabel.text,
				searchDirection: 'right'
			})
		);
	}

	const previousBalance = parseAmount(previousBalanceValue.text.str);

	if (previousBalance.isErr) {
		return Result.err(
			ParseStatementError.InvalidValue(`Invalid previous balance: ${previousBalance.error}`, {
				pageNumber: previousBalanceValue.pageNumber
			})
		);
	}

	const accountActivityLabel = statement.navigator.find(/Account Activity/);

	if (!accountActivityLabel) {
		return Result.err(ParseStatementError.MissingLabel('Missing "Account Activity" label'));
	}

	const totalPaymentsLabel = accountActivityLabel.findDown(/Total payments for this period/, {
		alignment: 'left'
	});

	if (!totalPaymentsLabel) {
		return Result.err(
			ParseStatementError.MissingLabel('Missing "Total payments for this period" label', {
				pageNumber: accountActivityLabel.pageNumber
			})
		);
	}

	const totalPaymentsValue = totalPaymentsLabel.findRight(MONEY_PATTERN);

	if (!totalPaymentsValue) {
		return Result.err(
			ParseStatementError.MissingValue('Missing total payments value', {
				searchFromText: totalPaymentsLabel.text,
				searchDirection: 'right'
			})
		);
	}

	const totalPayments = parseAmount(totalPaymentsValue.text.str);

	if (totalPayments.isErr) {
		return Result.err(
			ParseStatementError.InvalidValue(`Invalid total payments: ${totalPayments.error}`, {
				pageNumber: totalPaymentsValue.pageNumber
			})
		);
	}

	const totalTransactionsLabel = totalPaymentsLabel.findDown(
		/Total charges, credits, and returns for this period/,
		{
			alignment: 'left'
		}
	);

	if (!totalTransactionsLabel) {
		return Result.err(
			ParseStatementError.MissingLabel(
				'Missing "Total charges, credits, and returns for this period" label',
				{
					pageNumber: totalPaymentsLabel.pageNumber
				}
			)
		);
	}

	const totalTransactionsValue = totalTransactionsLabel.findRight(MONEY_PATTERN);

	if (!totalTransactionsValue) {
		return Result.err(
			ParseStatementError.MissingValue('Missing total transactions value', {
				searchFromText: totalTransactionsLabel.text,
				searchDirection: 'right'
			})
		);
	}

	const totalTransactions = parseAmount(totalTransactionsValue.text.str);

	if (totalTransactions.isErr) {
		return Result.err(
			ParseStatementError.InvalidValue(`Invalid total transactions: ${totalTransactions.error}`, {
				pageNumber: totalTransactionsValue.pageNumber
			})
		);
	}

	const totalDailyCashLabel = totalTransactionsLabel.findDown(/Total Daily Cash to account/, {
		alignment: 'left'
	});

	if (!totalDailyCashLabel) {
		return Result.err(
			ParseStatementError.MissingLabel('Missing "Total Daily Cash to account" label', {
				pageNumber: totalTransactionsLabel.pageNumber
			})
		);
	}

	const totalDailyCashValue = totalDailyCashLabel.findRight(MONEY_PATTERN);

	if (!totalDailyCashValue) {
		return Result.err(
			ParseStatementError.MissingValue('Missing total daily cash value', {
				searchFromText: totalDailyCashLabel.text,
				searchDirection: 'right'
			})
		);
	}

	const totalDailyCash = parseAmount(totalDailyCashValue.text.str);

	const totalInterestChargedLabel = totalDailyCashLabel.findDown(/Total interest for this month/, {
		alignment: 'left'
	});

	if (!totalInterestChargedLabel) {
		return Result.err(
			ParseStatementError.MissingLabel('Missing "Total interest for this month" label', {
				pageNumber: totalDailyCashLabel.pageNumber
			})
		);
	}

	const totalInterestChargedValue = totalInterestChargedLabel.findRight(MONEY_PATTERN);
	if (!totalInterestChargedValue) {
		return Result.err(
			ParseStatementError.MissingValue('Missing total interest charged value', {
				searchFromText: totalInterestChargedLabel.text,
				searchDirection: 'right'
			})
		);
	}

	const totalInterestCharged = parseAmount(totalInterestChargedValue.text.str);

	if (totalInterestCharged.isErr) {
		return Result.err(
			ParseStatementError.InvalidValue(
				`Invalid total interest charged: ${totalInterestCharged.error}`,
				{
					pageNumber: totalInterestChargedValue.pageNumber
				}
			)
		);
	}

	if (totalDailyCash.isErr) {
		return Result.err(
			ParseStatementError.InvalidValue(`Invalid total daily cash: ${totalDailyCash.error}`, {
				pageNumber: totalDailyCashValue.pageNumber
			})
		);
	}

	return Result.ok(
		new StatementSummary(
			startDate,
			closingDate,
			paymentDueDate,
			-previousBalance.value,
			-totalPayments.value,
			-totalTransactions.value,
			totalDailyCash.value,
			-totalInterestCharged.value,
			-yourBalance.value
		)
	);
}

export class StatementSummary extends StatementMetadata {
	startDate: DateTime;
	paymentDueDate: DateTime;
	previousBalance: number;
	totalPayments: number;
	totalTransactions: number;
	totalDailyCash: number;
	interestCharged: number;
	totalBalance: number;

	constructor(
		startDate: DateTime,
		closingDate: DateTime,
		paymentDueDate: DateTime,
		previousBalance: number,
		totalPayments: number,
		totalTransactions: number,
		totalDailyCash: number,
		interestCharged: number,
		totalBalance: number
	) {
		super(closingDate, 'credit', 'apple-card', 'Apple Card');
		this.startDate = startDate;
		this.closingDate = closingDate;
		this.paymentDueDate = paymentDueDate;
		this.previousBalance = previousBalance;
		this.totalPayments = totalPayments;
		this.totalTransactions = totalTransactions;
		this.totalDailyCash = totalDailyCash;
		this.interestCharged = interestCharged;
		this.totalBalance = totalBalance;
	}
}
