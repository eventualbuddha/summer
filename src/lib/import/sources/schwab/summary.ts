import { Result } from '@badrap/result';
import { DateTime, Interval } from 'luxon';
import { StatementMetadata, type AccountType } from '../../StatementMetadata';
import { ParseStatementError } from '../../parse/errors';
import { parseAmount } from '../../parse/money';
import type { Page } from '../../statement/page';

export class StatementSummary extends StatementMetadata {
	period: Interval;
	beginningBalance: number;
	depositsAndCredits: number;
	interestPaid: number;
	withdrawalsAndDebits: number;
	otherFees: number;
	endingBalance: number;

	constructor(
		period: Interval<true>,
		accountType: AccountType,
		account: string,
		accountName: string,
		beginningBalance: number,
		depositsAndCredits: number,
		interestPaid: number,
		withdrawalsAndDebits: number,
		otherFees: number,
		endingBalance: number
	) {
		super(period.end, accountType, account, accountName);
		this.period = period;
		this.beginningBalance = beginningBalance;
		this.depositsAndCredits = depositsAndCredits;
		this.interestPaid = interestPaid;
		this.withdrawalsAndDebits = withdrawalsAndDebits;
		this.otherFees = otherFees;
		this.endingBalance = endingBalance;
	}
}

export function parseStatementSummary(page: Page): Result<StatementSummary, ParseStatementError> {
	const summary: Partial<StatementSummary> = {};
	const parsePeriodResult = parseStatementPeriod(page);

	if (parsePeriodResult.isErr) {
		return Result.err(
			ParseStatementError.InvalidValue(
				`Invalid statement period: ${parsePeriodResult.error.message}`,
				{ pageNumber: page.pageNumber, cause: parsePeriodResult.error }
			)
		);
	}

	summary.period = parsePeriodResult.value;

	const accountNumberLabel = page.navigator.find(/^Account Number: (.+)$/);

	if (!accountNumberLabel) {
		return Result.err(
			ParseStatementError.MissingValue('Account number not found', { pageNumber: page.pageNumber })
		);
	}

	summary.account = accountNumberLabel.text.str.replace('Account Number: ', '');

	const accountNameLabel = accountNumberLabel.findLeft(/\w{5,}/);

	if (!accountNameLabel) {
		return Result.err(
			ParseStatementError.MissingValue('Account name not found', {
				pageNumber: page.pageNumber,
				searchFromText: accountNumberLabel.text,
				searchDirection: 'left'
			})
		);
	}

	summary.accountName = accountNameLabel.text.str;

	const summaryLabel = page.navigator.find('Summary');

	if (!summaryLabel) {
		return Result.err(
			ParseStatementError.MissingLabel('"Summary" label not found', { pageNumber: page.pageNumber })
		);
	}

	const amountLabel = summaryLabel.findRight('Amount');

	if (!amountLabel) {
		return Result.err(
			ParseStatementError.MissingLabel('"Amount" label not found', { pageNumber: page.pageNumber })
		);
	}

	const beginningBalanceLabel = summaryLabel.findDown('Beginning Balance', {
		alignment: 'left',
		maxGap: 20
	});

	if (!beginningBalanceLabel) {
		return Result.err(
			ParseStatementError.MissingLabel('"Beginning Balance" label not found', {
				pageNumber: page.pageNumber,
				searchFromText: summaryLabel.text,
				searchDirection: 'down'
			})
		);
	}

	const entries = beginningBalanceLabel?.findTable(
		[
			'Beginning Balance',
			'Deposits and Credits',
			'Interest Paid',
			'Withdrawals and Other Debits',
			'Other Fees',
			'Ending Balance'
		],
		[/\d+/, /\d+/, /\d+/, /\d+/, /\d+/, /\d+/],
		{ labelAlignment: 'left', valueAlignment: 'right', maxGap: 20 }
	);

	if (!entries) {
		return Result.err(
			ParseStatementError.MissingLabel('Summary table not found', { pageNumber: page.pageNumber })
		);
	}

	for (const [key, [, valueText]] of [
		['beginningBalance', entries[0]],
		['depositsAndCredits', entries[1]],
		['interestPaid', entries[2]],
		['withdrawalsAndDebits', entries[3]],
		['otherFees', entries[4]],
		['endingBalance', entries[5]]
	] as const) {
		const parseResult = parseAmount(valueText.text.str);

		if (parseResult.isErr) {
			return Result.err(
				ParseStatementError.InvalidValue(`Failed to parse ${key}: ${parseResult.error.message}`, {
					pageNumber: page.pageNumber,
					cause: parseResult.error
				})
			);
		}

		summary[key] = parseResult.value;
	}

	if (!summary.period || !summary.period.isValid) {
		return Result.err(
			ParseStatementError.InvalidValue(`"Period" not found or is invalid`, {
				pageNumber: page.pageNumber
			})
		);
	}

	if (!summary.account) {
		return Result.err(
			ParseStatementError.MissingLabel(`"Account" not found`, {
				pageNumber: page.pageNumber
			})
		);
	}

	if (!summary.accountName) {
		return Result.err(
			ParseStatementError.MissingLabel(`"Account Name" not found`, { pageNumber: page.pageNumber })
		);
	}

	if (typeof summary.beginningBalance === 'undefined') {
		return Result.err(
			ParseStatementError.MissingLabel(`"Beginning Balance" not found`, {
				pageNumber: page.pageNumber
			})
		);
	}

	if (typeof summary.depositsAndCredits === 'undefined') {
		return Result.err(
			ParseStatementError.MissingLabel(`"Deposits and Credits" not found`, {
				pageNumber: page.pageNumber
			})
		);
	}

	if (typeof summary.interestPaid === 'undefined') {
		return Result.err(
			ParseStatementError.MissingLabel(`"Interest Paid" not found`, { pageNumber: page.pageNumber })
		);
	}

	if (typeof summary.withdrawalsAndDebits === 'undefined') {
		return Result.err(
			ParseStatementError.MissingLabel(`"Withdrawals and Debits" not found`, {
				pageNumber: page.pageNumber
			})
		);
	}

	if (typeof summary.otherFees === 'undefined') {
		return Result.err(
			ParseStatementError.MissingLabel(`"Other Fees" not found`, { pageNumber: page.pageNumber })
		);
	}

	if (typeof summary.endingBalance === 'undefined') {
		return Result.err(
			ParseStatementError.MissingLabel(`"Ending Balance" not found`, {
				pageNumber: page.pageNumber
			})
		);
	}

	return Result.ok(
		new StatementSummary(
			summary.period,
			'banking',
			summary.account,
			summary.accountName,
			summary.beginningBalance,
			summary.depositsAndCredits,
			summary.interestPaid,
			summary.withdrawalsAndDebits,
			summary.otherFees,
			summary.endingBalance
		)
	);
}

export function parseStatementPeriodLines(
	firstLine: string,
	secondLine?: string
): Result<Interval, ParseStatementError> {
	if (secondLine) {
		const startMatch = firstLine.match(/^(\w+) (\d+), (\d+) to$/);

		if (!startMatch) {
			return Result.err(
				ParseStatementError.InvalidValue(
					`First statement period line does not match expected format: ${firstLine}`
				)
			);
		}

		const start = DateTime.fromFormat(
			`${startMatch[1]} ${startMatch[2]?.padStart(2, '0')}, ${startMatch[3]}`,
			'MMMM dd, yyyy'
		);

		const endMatch = secondLine.match(/^(\w+) (\d+), (\d+)$/);

		if (!endMatch) {
			return Result.err(
				ParseStatementError.InvalidValue(
					`Second statement period line does not match expected format: ${secondLine}`
				)
			);
		}

		const end = DateTime.fromFormat(
			`${endMatch[1]} ${endMatch[2]?.padStart(2, '0')}, ${endMatch[3]}`,
			'MMMM dd, yyyy'
		);

		return Result.ok(Interval.fromDateTimes(start, end));
	} else {
		const match = firstLine.match(/^(\w+) (\d+)-(\d+), (\d+)$/);

		if (!match) {
			return Result.err(
				ParseStatementError.InvalidValue(
					`First statement period line does not match expected format: ${firstLine}`
				)
			);
		}

		const start = DateTime.fromFormat(
			`${match[1]} ${match[2]?.padStart(2, '0')}, ${match[4]}`,
			'MMMM dd, yyyy'
		);
		const end = DateTime.fromFormat(
			`${match[1]} ${match[3]?.padStart(2, '0')}, ${match[4]}`,
			'MMMM dd, yyyy'
		);

		return Result.ok(Interval.fromDateTimes(start, end));
	}
}

function parseStatementPeriod(page: Page): Result<Interval<true>, ParseStatementError> {
	const statementPeriodLabel = page.navigator.find('Statement Period');

	if (!statementPeriodLabel) {
		return Result.err(
			ParseStatementError.MissingLabel('Statement period label not found', {
				pageNumber: page.pageNumber
			})
		);
	}

	const firstStatementPeriodLine = statementPeriodLabel.findDown(/.+/, {
		alignment: 'left',
		maxGap: 5
	});

	if (!firstStatementPeriodLine) {
		return Result.err(
			ParseStatementError.MissingValue('First statement period line not found', {
				pageNumber: page.pageNumber,
				searchFromText: statementPeriodLabel.text,
				searchDirection: 'down'
			})
		);
	}

	const secondStatementPeriodLine = firstStatementPeriodLine.findDown(/.+/, {
		alignment: 'left',
		maxGap: 5
	});

	return parseStatementPeriodLines(
		firstStatementPeriodLine.text.str,
		secondStatementPeriodLine?.text.str
	);
}
