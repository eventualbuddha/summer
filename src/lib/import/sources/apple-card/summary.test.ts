import { expect, test } from 'bun:test';
import { apple } from '../../../../../tests/fixtures';
import { parseStatementSummary } from './summary';

test('parse summary', () => {
	expect(parseStatementSummary(apple.statement())).toMatchSnapshot();
});
