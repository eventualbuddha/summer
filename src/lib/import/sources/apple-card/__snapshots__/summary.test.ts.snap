// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`parse summary 1`] = `
t {
  "isErr": false,
  "isOk": true,
  "value": StatementSummary {
    "account": "apple-card",
    "accountName": "Apple Card",
    "accountType": "credit",
    "closingDate": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 31,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 1,
        "second": 0,
        "year": 2025,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1738281600000,
      "weekData": null,
    },
    "interestCharged": -0,
    "paymentDueDate": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 28,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 2,
        "second": 0,
        "year": 2025,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1740700800000,
      "weekData": null,
    },
    "previousBalance": -85445,
    "startDate": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 1,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 1,
        "second": 0,
        "year": 2025,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1735689600000,
      "weekData": null,
    },
    "totalBalance": -97920,
    "totalDailyCash": 1377,
    "totalPayments": 85445,
    "totalTransactions": -97920,
  },
}
`;
